frb_rust/src/ffi/web.rs

```diff
@@ -0,0 +1,388 @@
+use super::DartAbi;
+use super::IntoDart;
+use super::MessagePort;
+pub use js_sys;
+pub use js_sys::Array as JsArray;
+use js_sys::*;
+use std::iter::FromIterator;
+pub use wasm_bindgen;
+pub use wasm_bindgen::closure::Closure;
+pub use wasm_bindgen::prelude::*;
+pub use wasm_bindgen::JsCast;
+use web_sys::BroadcastChannel;
+use web_sys::{DedicatedWorkerGlobalScope, Worker};
+
+pub trait IntoDartExceptPrimitive: IntoDart {}
+impl IntoDartExceptPrimitive for JsValue {}
+impl IntoDartExceptPrimitive for String {}
+impl<T: IntoDart> IntoDartExceptPrimitive for Option<T> {}
+
+mod pointer {
+    use std::ptr::NonNull;
+    use wasm_bindgen::convert::*;
+    use wasm_bindgen::describe::*;
+    use wasm_bindgen::JsValue;
+
+    /// A non-nullable pointer wrapping a [`NonNull`].
+    /// Represented in JS as an unsigned integer.
+    ///
+    /// Designed as a replacement for [`Option<*mut T>`].
+    ///
+    /// Usage | Parameter | Return
+    /// ----- | --------- | ------
+    /// `T` | Yes | Yes
+    /// `Option<T>` | Yes | Yes
+    /// `&T` | No | No
+    /// `&mut T` | No | No
+    pub struct Pointer<T>(NonNull<T>);
+
+    impl<T> Pointer<T> {
+        #[inline]
+        pub const fn as_mut(self) -> *mut T {
+            self.0.as_ptr() as _
+        }
+        #[inline]
+        pub fn from_js(value: &JsValue) -> Option<Self> {
+            NonNull::new(value.unchecked_into_f64() as u32 as *mut T).map(Self)
+        }
+    }
+
+    // HACK: This is wasm_bindgen's implementation detail, so it is not expected to
+    // be stable. In the future when `Option<*mut T>` can be exported to JS,
+    // this struct should be deprecated.
+    impl<T> WasmDescribe for Pointer<T> {
+        fn describe() {
+            JsValue::describe()
+        }
+    }
+
+    impl<T> FromWasmAbi for Pointer<T> {
+        type Abi = <JsValue as FromWasmAbi>::Abi;
+
+        unsafe fn from_abi(js: Self::Abi) -> Self {
+            let js = JsValue::from_abi(js);
+            Self::from_js(&js).expect("Attempted to retrieve a nullptr")
+        }
+    }
+
+    impl<T> OptionFromWasmAbi for Pointer<T> {
+        fn is_none(abi: &Self::Abi) -> bool {
+            *abi == 0
+        }
+    }
+
+    impl<T> IntoWasmAbi for Pointer<T> {
+        type Abi = u32;
+
+        fn into_abi(self) -> Self::Abi {
+            self.0.as_ptr() as _
```

Comment:
So our pointer must be within 32bits? That sounds impossible given today's addresses are virtualized and in 64-bit address space.

At least, maybe we should do a sanity check here, ensuring no overflow happens? (e.g. https://docs.rs/checked_int_cast/1.0.0/checked_int_cast/ or I guess rust already has some function like that)

Reply:
Unsound conversion, see above.